### ISSUE TITLE
[csrng/edn] instantiate lock-up?

### ISSUE THREAD

**Issue by tjaychen** [lowRISC/opentitan#14409]
I'm running into what appears to be a lock-up between otbn / csrng. 

![image](https://user-images.githubusercontent.com/1026250/185725176-86e85551-acfc-4c30-9c71-cdbba56039f9.png)

I've attached the waves above, but generally speaking this is the scenario.

1. csrng and edn0/edn1 are both activated
2. edn1 specifically makes an instantiation request with csrng
3. Before the instantiation request is completed and ack'd, the csrng is disabled
4. csrng and edn0/edn1 are re-enabled.  edn1 sends another instantiation request
5. This instantiation request is never serverd.

As far as I can tell, this is what's happening.
At step 3, when csrng is disabled, the queue of who has requested csrng service is wiped. So the fact that edn1 made a request is forgotten. 
At step 4 when edn1 makes a new request, it is not handled because it looks like `csrng_cmd_stage` did not reset its state due to the disable. So as a result no "new" request is issued and the state remains pending for the previous ack that never comes.

As a result, edn1 consumers are starved, and the distribution never completes.

**Comment by tjaychen** [comment]
I will push the modifications need to reproduce this soon.
It was hidden earlier because `entropy_src` was misconfigured in the test, see #13959.  I thought I had fixed it, but I think through some kind of merge conflict it crept back it. 

This misconfiguration generates a lot of "fake" entropy and allows all the steps to complete way before they should. 

Still, there could be a user error here, so let me know if something seems off to you.

**Comment by martin-lueker** [comment]
Is there some reason CSRNG is being disabled in step 3?

**Comment by andreaskurth** [comment]
CC @vogelpi

**Comment by andreaskurth** [comment]
> Is there some reason CSRNG is being disabled in step 3?

Yes, please see https://github.com/lowRISC/opentitan/issues/13561#issuecomment-1180553769

**Comment by andreaskurth** [comment]
I am assuming steps 1-4 are performed as the entropy components are first configured for boot mode and then for auto mode, so the procedure in general does not seem odd to me. However, the *disable before request completed* part seems odd.

You wrote "Before the instantiation request is completed and ack'd, the csrng is disabled"; did you check the `crsng_req_ready` and `csrng_rsp_ack` signals to confirm this?

Which test are you running? https://github.com/lowRISC/opentitan/pull/14410 mentions `csrng_aes_halt`, but I could not find test, and you mentioned OTBN and not AES in this issue. (I am asking because I would like to reproduce, so I can give more substantiated comments.)

**Comment by tjaychen** [comment]
@martin-lueker  / @andreaskurth / @mwbranstad 
yes that's right, the disable is happening because the test is trying to switch from `boot` mode entropy to fips entropy with auto mode enabled in `edn`.   So as a result every single block in the entropy_complex is being disabled and re-enabled. 

@andreaskurth 
if you look at the waveform, you can see the `csrng` command ready state. It has not fully ack'd and the disable is triggered. 
If you look at #14410 , just pull down that PR and run the entropy request test.  As I mentioned in another PR, the entropy request test currently sets the health test window count to 0.  Because of this, it triggers a bunch of unintended side effects, one of which is that nothing is health checked, the other is that entropy "appears" extremely fast.  So this issue doesn't end up showing up because `csrng` is already acked.

@mwbranstad 
i think that's true.  I haven't looked at `edn` in detail, but we might need some kind of extra check that when `edn` is disabled we aren't in the middle of sending entropy back to the end points and thus cause a failure.  

This specific failure though isn't because of the `edn`.  But it appears to be inside the `cmd_stage` fsm, which doesn't seem to be accounting for `csrng` enable/disable.  The other point you made about the `edn` is still possible though. But I think it would be better to re-create that test at block level.

**Comment by mwbranstad** [comment]
@tjaychen regarding cmd_stage_fsm - I believe @senelson7 has a pending fix for coverage that also includes checking each state for enable, such that part of the problem in this issue may go away if that PR is approved.

For stalling the edn endpoints, it is not clear what the best implementation might be. It could be in EDN but I think we may need to instruct the endpoint to stop requesting, then wait for an endpoint status that says it is quiesced. Once in that state, we can then safely disable the edn/csrng/es modules. Once possible solution anyway.

**Comment by tjaychen** [comment]
@mwbranstad do you mean @senelson7 fix also adds enable handling in the fsm..? 
The fsm i'm referring to is [here](https://github.com/lowRISC/opentitan/blob/master/hw/ip/csrng/rtl/csrng_cmd_stage.sv#L255), i don't immediately see anything obvious in the `CmdAck` state that [handles disable](https://github.com/lowRISC/opentitan/blob/master/hw/ip/csrng/rtl/csrng_cmd_stage.sv#L300), and that was my assumption as to why things wedged.

If it's okay with you, I can do some local trials to just transition everything back to `Idle` state and see if it clears things up.

I agree with you on `edn`.  I think if we do want to handle it, we probably need to look at the fsm interfacing with each endpoint (is it [this](https://github.com/lowRISC/opentitan/blob/master/hw/ip/edn/rtl/edn_ack_sm.sv#L52)?).  It actually looks like this interface is very simple so maybe we are fine already.

**Comment by mwbranstad** [comment]
Yes @senelson7 fix was added to to be consistent with the main sm. He added it mostly for coverage, but it also has a functional component as you would imagine.
I think it is being blocked by how forces should be done. I think it will be addressed in tomorrow's meeting possible.

**Comment by tjaychen** [comment]
sounds good. Then let me add a local hack to move on for now. The `edn` one i think probably makes more sense to test at block level. Ie, just randomly disable during the middle of a lot of activity and see if anything locks up.  What do you think?

**Comment by tjaychen** [comment]
#14028 will likely address the issue presented in this PR.

**Comment by tjaychen** [comment]
this has been addressed and can be closed.
